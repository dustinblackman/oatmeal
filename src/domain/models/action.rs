use super::AcceptType;
use super::BackendPrompt;
use super::EditorContext;
use super::Message;

pub enum Action {
    AcceptCodeBlock(Option<EditorContext>, String, AcceptType),
    BackendAbort(),
    BackendRequest(BackendPrompt),
    CopyMessages(Vec<Message>),
    EditPromptBegin(),
}
