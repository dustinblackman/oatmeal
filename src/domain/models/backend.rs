#[cfg(test)]
#[path = "backend_test.rs"]
mod tests;

use anyhow::Result;
use async_trait::async_trait;
use tokio::sync::mpsc;

use super::Author;
use super::EditorContext;
use super::Event;

pub struct BackendPrompt {
    pub text: String,
    pub backend_context: String,
}

impl BackendPrompt {
    pub fn new(text: String, backend_context: String) -> BackendPrompt {
        return BackendPrompt {
            text,
            backend_context,
        };
    }

    pub fn append_system_prompt(&mut self, editor_context: &Option<EditorContext>) {
        if let Some(context) = editor_context {
            let lang = &context.language;
            let code = &context.code;

            let system_prompt = format!(". The coding language is {lang}. Return results in markdown, add language to code blocks.");
            self.text += &system_prompt;

            if !code.is_empty() {
                let code_prompt = format!("The code is the following: {code}");
                self.text += &code_prompt;
            }
        } else {
            self.text += ". Return results in markdown, add language to code blocks."
        }
    }
}

#[derive(Clone)]
pub struct BackendResponse {
    pub author: Author,
    pub text: String,
    pub done: bool,
    pub context: Option<String>,
}

#[async_trait]
pub trait Backend {
    /// Used at startup to verify all configurations are available to work with
    /// the backend.
    async fn health_check(&self) -> Result<()>;

    /// Called when using the `/modellist` slash commands to provide all
    /// available models for the backend.
    async fn list_models<'a>(&'a self) -> Result<Vec<String>>;

    /// Requests completions from the backend. Completion results may be
    /// streamed back to the UI by passing each should through a channel.
    ///
    /// Upon receiving all results, a final `done` boolean
    /// is provided as the last message to the channel.
    ///
    /// In order for a backend to maintain history, a context array must be
    /// provided by the backend. This can be passed alongside the `done`
    /// boolean, and will be provided on the next prompt to the backend.
    async fn get_completion<'a>(
        &self,
        prompt: BackendPrompt,
        tx: &'a mpsc::UnboundedSender<Event>,
    ) -> Result<()>;
}
