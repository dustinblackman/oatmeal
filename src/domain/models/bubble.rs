#[cfg(test)]
#[path = "bubble_test.rs"]
mod tests;

use once_cell::sync::Lazy;
use ratatui::style::Color;
use ratatui::style::Style;
use ratatui::text::Line;
use ratatui::text::Span;
use syntect::easy::HighlightLines;
use syntect::highlighting::Theme;
use syntect::parsing::SyntaxReference;
use syntect::parsing::SyntaxSet;

use super::Author;
use super::Message;
use super::MessageType;
use crate::domain::services::Syntaxes;

// TODO include this in refactors, it shouldn't be here.
static SYNTAX: Lazy<SyntaxSet> = Lazy::new(Syntaxes::load);

#[derive(PartialEq, Eq)]
pub enum BubbleAlignment {
    Left,
    Right,
}

pub struct Bubble {
    message: Message,
}

// TODO this entire model has gotten out of hand. Refactor.
impl<'a> Bubble {
    pub fn new(message: Message) -> Bubble {
        return Bubble { message };
    }

    pub fn as_lines(
        &self,
        alignment: BubbleAlignment,
        theme: &Theme,
        window_max_width: u16,
        total_codeblock_counter: usize,
    ) -> Vec<Line<'a>> {
        // Lazy defaults
        let syntax = get_syntax("text");
        let mut highlight = HighlightLines::new(syntax, theme);

        // Add a minimum 4% of padding on the side.
        let min_bubble_padding_length = ((window_max_width as f32 * 0.04).ceil()) as usize;

        // left border + left padding + (text, not counted) + right padding + right
        // border + scrollbar. And then minimum bubble padding.
        let line_border_width = 5 + min_bubble_padding_length;

        let message_lines = self
            .message
            .as_string_lines(window_max_width - line_border_width as u16);

        let username = &self.message.author_formatted;
        let mut max_line_length = message_lines
            .iter()
            .map(|line| {
                return line.len();
            })
            .max()
            .unwrap();
        if max_line_length < username.len() {
            max_line_length = username.len();
        }

        let mut in_codeblock = false;
        let mut codeblock_count = 0;
        let mut lines: Vec<Line> = vec![];

        for line in message_lines {
            let (formatted_line, mut spans) = self.format_line(line.to_string(), max_line_length);
            let bubble_padding = [" "]
                .repeat(window_max_width as usize - formatted_line.len())
                .join("");

            if in_codeblock {
                let highlighted_spans: Vec<Span> = highlight
                    .highlight_line(&line, &SYNTAX)
                    .unwrap()
                    .iter()
                    .map(|segment| {
                        let (style, content) = segment;

                        return Span::styled(
                            content.to_string(),
                            Style {
                                fg: translate_colour(style.foreground),
                                ..Style::default()
                            },
                        );
                    })
                    .collect();

                spans = self
                    .format_spans(line.to_string(), max_line_length, highlighted_spans)
                    .1;
            }

            if line.trim().starts_with("```") {
                let lang = line.trim().replace("```", "");
                let syntax = get_syntax(&lang);
                if syntax.name.to_lowercase() != "plain text" {
                    highlight = HighlightLines::new(syntax, theme);
                    in_codeblock = true;

                    codeblock_count += 1;
                    spans = self
                        .format_spans(
                            format!("{line} ({})", total_codeblock_counter + codeblock_count),
                            max_line_length,
                            vec![
                                Span::from(line),
                                Span::styled(
                                    format!(" ({})", total_codeblock_counter + codeblock_count),
                                    Style {
                                        fg: Some(Color::White),
                                        ..Style::default()
                                    },
                                ),
                            ],
                        )
                        .1;
                } else {
                    in_codeblock = false;
                }
            }

            if alignment == BubbleAlignment::Left {
                spans.push(Span::from(bubble_padding));
                lines.push(Line::from(spans));
            } else {
                let mut res = vec![Span::from(bubble_padding)];
                res.extend(spans);
                lines.push(Line::from(res));
            }
        }

        // Add 2 for the vertical bars.
        let inner_bar = ["─"].repeat(max_line_length + 2).join("");
        let top_left_border = "╭";
        let mut top_bar = format!("{top_left_border}{inner_bar}╮");
        let bottom_bar = format!("╰{inner_bar}╯");
        let bar_bubble_padding = [" "]
            // TODO WTF is 8?
            .repeat(window_max_width as usize - max_line_length - 8)
            .join("");

        if alignment == BubbleAlignment::Left {
            let top_replace = ["─"].repeat(username.len()).join("");
            top_bar = top_bar.replace(
                format!("{top_left_border}{top_replace}").as_str(),
                format!("{top_left_border}{username}").as_str(),
            );

            let mut res = vec![self.highlight_line(format!("{top_bar}{bar_bubble_padding}"))];
            res.extend(lines);
            res.push(self.highlight_line(format!("{bottom_bar}{bar_bubble_padding}")));
            return res;
        } else {
            let top_replace = ["─"].repeat(username.len()).join("");
            top_bar = top_bar.replace(
                format!("{top_left_border}{top_replace}").as_str(),
                format!("{top_left_border}{username}").as_str(),
            );

            let mut res = vec![self.highlight_line(format!("{bar_bubble_padding}{top_bar}"))];
            res.extend(lines);
            res.push(self.highlight_line(format!("{bar_bubble_padding}{bottom_bar}")));
            return res;
        }
    }

    fn format_spans(
        &self,
        line: String,
        max_line_length: usize,
        mut spans: Vec<Span<'a>>,
    ) -> (String, Vec<Span<'a>>) {
        let fill = [" "].repeat(max_line_length - line.len()).join("");
        let formatted_line = format!("│ {line}{fill} │");

        let mut spans_res = vec![self.highlight_span("│ ".to_string())];
        spans_res.append(&mut spans);
        spans_res.push(self.highlight_span(format!("{fill} │").to_string()));
        return (formatted_line, spans_res);
    }

    fn format_line(&self, line: String, max_line_length: usize) -> (String, Vec<Span<'a>>) {
        return self.format_spans(
            line.to_string(),
            max_line_length,
            vec![Span::from(line.clone())],
        );
    }

    fn highlight_span(&self, text: String) -> Span<'a> {
        if self.message.message_type() == MessageType::Error {
            return Span::styled(
                text,
                Style {
                    fg: Some(Color::Red),
                    ..Style::default()
                },
            );
        } else if self.message.author == Author::Oatmeal {
            return Span::styled(
                text,
                Style {
                    fg: Some(Color::Rgb(138, 85, 63)), // Brown
                    ..Style::default()
                },
            );
        }

        return Span::from(text);
    }

    fn highlight_line(&self, text: String) -> Line<'a> {
        return Line::from(self.highlight_span(text));
    }
}

fn translate_colour(syntect_color: syntect::highlighting::Color) -> Option<Color> {
    match syntect_color {
        syntect::highlighting::Color { r, g, b, a } if a > 0 => return Some(Color::Rgb(r, g, b)),
        _ => return None,
    }
}

fn get_syntax(name: &str) -> &SyntaxReference {
    if let Some(syntax) = SYNTAX.find_syntax_by_extension(name) {
        return syntax;
    }

    if let Some(syntax) = SYNTAX.find_syntax_by_name(name) {
        return syntax;
    }

    if let Some(syntax) = SYNTAX.find_syntax_by_token(name) {
        return syntax;
    }

    return SYNTAX.find_syntax_plain_text();
}
