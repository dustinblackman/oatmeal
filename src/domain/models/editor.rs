#[cfg(test)]
#[path = "editor_test.rs"]
mod tests;

use std::fmt;

use anyhow::Result;
use async_trait::async_trait;
use strum::EnumIter;
use strum::EnumVariantNames;
use strum::IntoEnumIterator;

#[derive(Clone, Debug, PartialEq, Eq, EnumIter, EnumVariantNames, strum::Display)]
#[strum(serialize_all = "lowercase")]
pub enum EditorName {
    Neovim,
    Clipboard,
    None,
}

impl EditorName {
    pub fn parse(text: String) -> Option<EditorName> {
        return EditorName::iter().find(|e| return e.to_string() == text);
    }
}

#[derive(Debug, PartialEq)]
pub enum AcceptType {
    /// Append in editor where the cursor was last.
    Append,
    /// Replace selected code in editor.
    Replace,
}

impl fmt::Display for AcceptType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            AcceptType::Append => return write!(f, "append"),
            AcceptType::Replace => return write!(f, "replace"),
        }
    }
}

#[derive(Clone, Default)]
pub struct EditorContext {
    pub file_path: String,
    pub language: String,
    pub code: String,
    pub start_line: i64,
    pub end_line: Option<i64>,
}

impl EditorContext {
    pub fn format(&self) -> String {
        let file_path = &self.file_path;
        let language = &self.language;
        let code = &self.code;

        if code.is_empty() || self.end_line.is_none() {
            return format!("File: {file_path}");
        }

        return format!(
            r#"
File: {file_path}

```{language}
{code}
```
    "#
        )
        .trim()
        .to_string();
    }
}

#[async_trait]
pub trait Editor {
    /// Returns the name of the editor.
    fn name(&self) -> EditorName;

    /// Used at startup to verify all configurations are available to work with
    /// the editor.
    async fn health_check(&self) -> Result<()>;

    /// Used to provide context from the editor such as coding language, file,
    /// selected lines, and full code blocks.
    async fn get_context(&self) -> Result<Option<EditorContext>>;

    /// If required, clear_context is called when Oatmeal exits to do any
    /// necessary cleanup in the editor.
    async fn clear_context(&self) -> Result<()>;

    /// Sends accepted code blocks back to the editor, with `accept_type`
    /// defining if code blocks should replace existing code, or append to
    /// it.
    async fn send_codeblock<'a>(
        &self,
        context: EditorContext,
        codeblock: String,
        accept_type: AcceptType,
    ) -> Result<()>;
}

pub type EditorBox = Box<dyn Editor + Send + Sync>;
