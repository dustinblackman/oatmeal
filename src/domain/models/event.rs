use tui_textarea::Input;

use super::BackendResponse;
use super::Message;

pub enum Event {
    BackendMessage(Message),
    BackendPromptResponse(BackendResponse),
    KeyboardCharInput(Input),
    KeyboardCTRLC(),
    KeyboardCTRLO(),
    KeyboardCTRLR(),
    KeyboardEnter(),
    KeyboardPaste(String),
    UITick(),
    UIScrollDown(),
    UIScrollUp(),
    UIScrollPageDown(),
    UIScrollPageUp(),
}
