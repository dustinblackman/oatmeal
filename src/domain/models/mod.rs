mod action;
mod backend;
mod bubble;
mod editor;
mod event;
mod loading;
mod message;
mod session;
mod slash_commands;
mod textarea;

pub use action::*;
pub use backend::*;
pub use bubble::*;
pub use editor::*;
pub use event::*;
pub use loading::*;
pub use message::*;
pub use session::*;
pub use slash_commands::*;
pub use textarea::*;
