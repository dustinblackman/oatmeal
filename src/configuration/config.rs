#[cfg(test)]
#[path = "config_test.rs"]
mod tests;

use std::env;
use std::path;

use anyhow::bail;
use anyhow::Result;
use clap::ArgMatches;
use clap::Command;
use dashmap::DashMap;
use once_cell::sync::Lazy;
use strum::EnumIter;
use strum::EnumVariantNames;
use strum::IntoEnumIterator;
use tokio::fs;

use crate::domain::models::BackendName;
use crate::domain::models::EditorName;

static CONFIG: Lazy<DashMap<String, String>> = Lazy::new(DashMap::new);

#[derive(Clone, Copy, Eq, PartialEq, EnumIter, EnumVariantNames, strum::Display)]
#[strum(serialize_all = "kebab-case")]
pub enum ConfigKey {
    Backend,
    BackendHealthCheckTimeout,
    Editor,
    Model,
    ConfigFile,
    LangChainURL,
    OllamaURL,
    OpenAiToken,
    OpenAiURL,
    ClaudeToken,
    SessionID,
    Theme,
    ThemeFile,
    Username,
}

pub struct Config {}

impl Config {
    pub fn get(key: ConfigKey) -> String {
        if let Some(val) = CONFIG.get(&key.to_string()) {
            return val.to_string();
        }

        return "".to_string();
    }

    pub fn set(key: ConfigKey, value: &str) {
        CONFIG.insert(key.to_string(), value.to_string());
    }

    pub fn default(key: ConfigKey) -> String {
        if key == ConfigKey::Username {
            let mut user = env::var("USER").unwrap_or_else(|_| return "".to_string());
            if user.is_empty() {
                user = "User".to_string();
            }

            return user;
        }

        let default_backend = BackendName::Ollama.to_string();
        let default_editor = EditorName::Clipboard.to_string();

        let mut config_path = dirs::cache_dir().unwrap().join("oatmeal/config.toml");

        #[cfg(target_os = "macos")]
        {
            config_path =
                path::PathBuf::from(env::var("HOME").unwrap()).join(".config/oatmeal/config.toml");
        }

        #[cfg(target_os = "windows")]
        {
            config_path = dirs::cache_dir().unwrap().join("oatmeal/config.toml");
        }

        #[cfg(target_os = "linux")]
        {
            config_path = dirs::cache_dir().unwrap().join("oatmeal/config.toml");
            if !config_path.exists() {
                config_path = dirs::config_local_dir()
                    .unwrap()
                    .join("oatmeal/config.toml");
            }
        }

        let res = match key {
            ConfigKey::Backend => &default_backend,
            ConfigKey::BackendHealthCheckTimeout => "1000",
            ConfigKey::Editor => &default_editor,
            ConfigKey::Model => "",
            ConfigKey::LangChainURL => "http://localhost:8000",
            ConfigKey::OllamaURL => "http://localhost:11434",
            ConfigKey::OpenAiToken => "",
            ConfigKey::OpenAiURL => "https://api.openai.com",
            ConfigKey::ClaudeToken => "",
            ConfigKey::Theme => "base16-onedark",
            ConfigKey::ThemeFile => "",

            // Special
            ConfigKey::ConfigFile => config_path.to_str().unwrap(),
            ConfigKey::SessionID => "",
            ConfigKey::Username => "",
        };

        return res.to_string();
    }

    pub async fn load(cmd: Command, clap_arg_matches: Vec<&ArgMatches>) -> Result<()> {
        for key in ConfigKey::iter() {
            Config::set(key, &Config::default(key))
        }

        let mut config_file = Config::default(ConfigKey::ConfigFile);
        for matches in clap_arg_matches.as_slice() {
            if let Some(arg_config_file) =
                matches.get_one::<String>(&ConfigKey::ConfigFile.to_string())
            {
                config_file = arg_config_file.to_string();
            }
        }

        let config_path = path::PathBuf::from(config_file);
        if config_path.exists() {
            let toml_str = fs::read_to_string(config_path).await?;
            let doc = toml_str.parse::<toml_edit::Document>()?;

            for key in ConfigKey::iter() {
                if let Some(val) = doc.get(&key.to_string()) {
                    // Use clap value parsers to do validation.
                    let mut possible_values = vec![];
                    if let Some(arg) = cmd
                        .get_arguments()
                        .find(|e| return e.get_long().unwrap() == key.to_string())
                    {
                        if !arg.get_possible_values().is_empty() {
                            possible_values = arg
                                .get_possible_values()
                                .iter()
                                .map(|e| return e.get_name().to_string())
                                .collect::<Vec<String>>();
                        }
                    }

                    if let Some(val_int) = val.as_integer() {
                        Config::set(key, &val_int.to_string());
                    } else if let Some(val_str) = val.as_str() {
                        if val_str.is_empty() {
                            continue;
                        }
                        if !possible_values.is_empty()
                            && !possible_values.contains(&val_str.to_string())
                        {
                            bail!(format!("config.toml has an invalid value for key '{key}': {val_str}\nPossible values are: {}", possible_values.join(", ")));
                        }
                        Config::set(key, val_str);
                    }
                }
            }
        }

        for key in ConfigKey::iter() {
            for matches in clap_arg_matches.as_slice() {
                if let Ok(Some(val)) = matches.try_get_one::<String>(&key.to_string()) {
                    if val.is_empty() {
                        continue;
                    }
                    Config::set(key, val)
                }
            }
        }

        tracing::debug!(
            username = Config::get(ConfigKey::Username),
            backend = Config::get(ConfigKey::Backend),
            editor = Config::get(ConfigKey::Editor),
            model = Config::get(ConfigKey::Model),
            theme = Config::get(ConfigKey::Theme),
            theme_file = Config::get(ConfigKey::ThemeFile),
            "config"
        );

        return Ok(());
    }

    pub fn serialize_default(cmd: Command) -> String {
        let toml_str = ConfigKey::iter()
            .filter_map(|key| {
                if key == ConfigKey::SessionID || key == ConfigKey::ConfigFile {
                    return None;
                }

                if key == ConfigKey::Username {
                    return Some(
                        "# Your user name displayed in all chat bubbles.\n# username = \"\""
                            .to_string(),
                    );
                }

                let arg = cmd
                    .get_arguments()
                    .find(|e| return e.get_long().unwrap() == key.to_string())
                    .unwrap();

                let mut description = arg.get_help().unwrap().to_string();

                description = description
                    .split("[default:")
                    .next()
                    .unwrap()
                    .trim()
                    .to_string();

                if !arg.get_possible_values().is_empty() {
                    let possible_values = arg
                        .get_possible_values()
                        .iter()
                        .map(|e| return e.get_name())
                        .collect::<Vec<_>>()
                        .join(", ");
                    description = format!("{description} [possible values: {}]", possible_values);
                }

                let mut val = Config::default(key);
                if val.is_empty() {
                    val = format!("# {key} = \"\"");
                } else if val.parse::<i32>().is_ok() {
                    val = format!("{key} = {val}");
                } else {
                    val = format!("{key} = \"{val}\"");
                }

                return Some(format!("# {description}\n{val}"));
            })
            .collect::<Vec<String>>()
            .join("\n\n");

        return toml_str;
    }
}
