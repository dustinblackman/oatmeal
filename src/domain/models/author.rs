use std::fmt;

use serde::Deserialize;
use serde::Serialize;

use crate::configuration::Config;
use crate::configuration::ConfigKey;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum Author {
    User,
    Oatmeal,
    Model,
}

impl fmt::Display for Author {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let name = match self {
            Author::User => Config::get(ConfigKey::Username),
            Author::Oatmeal => String::from("Oatmeal"),
            Author::Model => Config::get(ConfigKey::Model),
        };
        return write!(f, "{}", name);
    }
}
