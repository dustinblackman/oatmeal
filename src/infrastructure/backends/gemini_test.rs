use anyhow::bail;
use anyhow::Result;
use test_utils::insta_snapshot;
use tokio::sync::mpsc;

use super::Config;
use super::Content;
use super::ContentParts;
use super::Gemini;
use super::Model;
use super::ModelListResponse;
use crate::configuration::ConfigKey;
use crate::domain::models::Author;
use crate::domain::models::Backend;
use crate::domain::models::BackendPrompt;
use crate::domain::models::BackendResponse;
use crate::domain::models::Event;

impl Gemini {
    fn with_url(url: String) -> Gemini {
        return Gemini {
            url,
            token: "abc".to_string(),
            timeout: "200".to_string(),
        };
    }
}

fn to_res(action: Option<Event>) -> Result<BackendResponse> {
    let act = match action.unwrap() {
        Event::BackendPromptResponse(res) => res,
        _ => bail!("Wrong type from recv"),
    };

    return Ok(act);
}

#[tokio::test]
async fn it_successfully_health_checks() {
    Config::set(ConfigKey::Model, "model-1");
    let mut server = mockito::Server::new();
    let mock = server
        .mock("GET", "/v1beta/model-1?key=abc")
        .with_status(200)
        .create();

    let backend = Gemini::with_url(server.url());
    let res = backend.health_check().await;

    assert!(res.is_ok());
    mock.assert();
}

#[tokio::test]
async fn it_successfully_health_checks_with_official_api() {
    Config::set(ConfigKey::Model, "models/gemini-pro");
    let token = match std::env::var("OATMEAL_GEMINI_TOKEN") {
        Ok(token) => token,
        Err(_) => {
            println!("There is no token in environment defined, skipping test");
            return;
        }
    };
    let backend = Gemini {
        url: "https://generativelanguage.googleapis.com".to_string(),
        token,
        timeout: "500".to_string(),
    };

    let res = backend.health_check().await;
    assert!(res.is_ok());
}

#[tokio::test]
async fn it_fails_health_checks() {
    Config::set(ConfigKey::Model, "model-1");
    let mut server = mockito::Server::new();
    let mock = server
        .mock("GET", "/v1beta/model-1?key=abc")
        .with_status(500)
        .create();

    let backend = Gemini::with_url(server.url());
    let res = backend.health_check().await;

    assert!(res.is_err());
    mock.assert();
}

#[tokio::test]
async fn it_lists_models() -> Result<()> {
    let body = serde_json::to_string(&ModelListResponse {
        models: vec![
            Model {
                name: "first".to_string(),
                supported_generation_methods: vec!["generateContent".to_string()],
            },
            Model {
                name: "second".to_string(),
                supported_generation_methods: vec!["generateContent".to_string()],
            },
        ],
    })?;

    let mut server = mockito::Server::new();
    let mock = server
        .mock("GET", "/v1beta/models?key=abc")
        .with_status(200)
        .with_body(body)
        .create();

    let backend = Gemini::with_url(server.url());
    let res = backend.list_models().await?;
    mock.assert();

    assert_eq!(res, vec!["first".to_string(), "second".to_string()]);

    return Ok(());
}

#[tokio::test]
async fn it_gets_completions() -> Result<()> {
    Config::set(ConfigKey::Model, "model-1");
    let body = [
        "[",
        "\"contents\": [{",
        "\"parts\": [{",
        "\"text\": \"Hello \"",
        "}]",
        "},",
        "{",
        "\"parts\": [{",
        "\"text\": \"World\"",
        "}]",
        "},",
        "{",
        "\"parts\": [{",
        "\"text\": \"\"",
        "}]",
        "}]",
        "]",
    ]
    .join("\n");
    let prompt = BackendPrompt {
        text: "Say hi to the world".to_string(),
        backend_context: serde_json::to_string(&vec![Content {
            role: "model".to_string(),
            parts: vec![ContentParts::Text("Hello".to_string())],
        }])?,
    };

    let mut server = mockito::Server::new();
    let mock = server
        .mock("POST", "/v1beta/model-1:streamGenerateContent?key=abc")
        .with_status(200)
        .with_body(body)
        .create();

    let (tx, mut rx) = mpsc::unbounded_channel::<Event>();

    let backend = Gemini::with_url(server.url());
    backend.get_completion(prompt, &tx).await?;

    mock.assert();

    let first_recv = to_res(rx.recv().await)?;
    let second_recv = to_res(rx.recv().await)?;
    let third_recv = to_res(rx.recv().await)?;

    assert_eq!(first_recv.author, Author::Model);
    assert_eq!(first_recv.text, "Hello ".to_string());
    assert!(!first_recv.done);
    assert_eq!(first_recv.context, None);

    assert_eq!(second_recv.author, Author::Model);
    assert_eq!(second_recv.text, "World".to_string());
    assert!(!second_recv.done);
    assert_eq!(second_recv.context, None);

    assert_eq!(third_recv.author, Author::Model);
    assert_eq!(third_recv.text, "".to_string());
    assert!(third_recv.done);
    insta_snapshot(|| {
        insta::assert_toml_snapshot!(third_recv.context);
    });

    return Ok(());
}
