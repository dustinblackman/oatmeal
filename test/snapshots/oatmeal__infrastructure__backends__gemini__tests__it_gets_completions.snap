---
source: src/infrastructure/backends/gemini_test.rs
expression: third_recv.context
---
'[{"role":"model","parts":[{"text":"Hello"}]},{"role":"user","parts":[{"text":"Say hi to the world"}]},{"role":"model","parts":[{"text":"Hello World"}]}]'
